### ISSUE TITLE
Improve the main menu

### ISSUE THREAD

**Issue by pierreporte** [OtterBrowser/otter-browser#1112]
The main menu is a very important element of the interface in the way that it is the entry of non-trivial features, that is, features that are not simply navigating from page to page. For this reason, this menu must be carefully designed to maximise the easiness to access features by providing a good choice of directly accessible features and to chose well the organisation of sub-menus.

I open this issue to discuss of what the main menu should look like. Yet am I aware that the main menu of Otter Browser is customisable, I believe that a good default configuration is capital. Even for power users, which are the target of our beloved programme, good design is always a good choice.

# References

## Firefox 3.7

![](https://wiki.mozilla.org/images/0/01/Fx-3.7-Tools-Menu-Phase-01.png)

## Firefox 4+

Normal:

![](https://addons.cdn.mozilla.net/user-media/previews/full/53/53794.png?modified=1297314009)

With extension (I think):

![](https://wiki.mozilla.org/images/a/a7/Firefox_Menu_button_2_column_win7_start_menu_style.png)

## Firefox Australis

![](http://i.imgur.com/Raa8XWr.jpg)

## Opera 11 or 12

No surprise, very similar to Otter Browser.

![](http://www.wired.com/images_blogs/business/2010/02/opera105menu.jpg)

## Opera 15+

![](http://help.opera.com/opera/Windows/1116/fr/images/UI/mainMenu.png)

## Chrome in february 2015

![](https://www.bugsfighter.com/wp-content/uploads/2015/02/google-chrome-menu-settings.png)

# My proposition

It is not necessarily complete but I tried to be exhaustive.

## Main menu

```
Menu > | [ Copy ] [ Cut ] [ Paste ]
       | Zoom  [ - ] [ 100% ] [ + ]
       | Tabs and windows         >
       | Page                     >
       | --------------------------
       | [ Bookmarks ]            >
       | [ History ]              >
       | [ Transfers ]            >
       | [ Notes ]                >
       | --------------------------
       | Options                  >
       | Help                     >
       | Quit
```

The c., c. & p. buttons are all on the same row. The particularity of the zoom buttons is that when clicking them, the menu doesn’t vanish. The middle button is used to display the current zoom level and to restore it to 100%.

I moved the “print” menu to the “page” submenu. The “bookmarks”, “history”, “transfers” and “notes” (plus the “addons” which was added after I worked on this issue) are quite special in the way that even if they have a sub-menu, clicking on them actually do something. Their action could be to display the side panel with the page corresponding to the menu (bookmarks panel, history panel, etc.) but that could be a setting for each one. the “toolbars” and “display menu bar” entries are moved into the “option” menu.

## Tabs and windows

```
Tabs and windows > | New tab
                   | New private tab
                   | Close tab
                   | Closed tabs      >
                   | ------------------
                   | New window
                   | New private window
                   | Close window
                   | Closed windows   >
                   | ------------------
                   | Sessions         >
```

I expanded the current menu with all features related to opening and closing tabs and windows. 

## Page

```
Page > | Open
       | Save
       | Print      >
       | ------------
       | Search
       | ------------
       | Full screen
       | ------------
       | View source
       | Inspect page
       | ------------
       | Style      >
       | Encoding   >
```

Quite everything related to the selections could be moved to the context menu. The navigation features already being in the tool bar and in the context menu, repeating them again is a bit too much. Speaking of the context menu, this is something we could also discuss about.

## Bookmarks, history, transfers and notes

```
Bookmarks > | Add bookmark
            | Manage bookmarsk
            | -----------------
            | Start page      >
            | (other folders)
            | (other bookmarks)
```

```
History > | Clear history
          | Manage history
```

```
Transfers > | New transfer
            | Manage transfers
```

```
Notes > | New note
        | Manage notes
```

All these are organised the same, except for the bookmarks where there is the list of bookmarks. the “manage X” entries open the about:X page.

## Options

```
Options > | Preferences
          | Toolbars     >
          | User agent   >
          | Reset defaults
          | --------------
          | Import       >
          | Export       >
          | --------------
          | Work offline
          | --------------
          | Show menu bar
```

In this menu are moved everything that has something to do with making modifications to Otter Browser.

## Help

```
Help > | About otter
       | Documentation
       | -----------------
       | Check updates
       | Diagnostic report
```

I removed the “about Qt” entry which is useless. The fact that Otter Browser uses Qt is not something so important that it needs to be advertised that much. In my opinion, a mention of it in the about:otter page is sufficient. I added a “documentation” entry: we absolutely need a place to explain how to do everything, show the keyboard shortcuts and the mouse gestures.

**Comment by Frenzie** [comment]
For the record, Firefox Australis is the ultimate exercise in unclarity. I've been using Firefox as essentially my main browser for a few years now and I can still barely find anything in there. Chrome is similarly awful, but not quite as bad. Opera/Presto and Otter are significantly superior. 

tl;dr I think several of the examples given are to be taken as ultimate examples of how *not* to do it. 

> I moved the “print” menu to the “page” submenu.

Although I use Ctrl + P, I think it's far too important to do that.

Now that I've got the negatives out of the way, I'd like to say that for the rest I either agree completely with your proposed changes or think your suggestions are worth some serious consideration. Thanks for supplying all of the screenshots and typing it all out in such a detailed manner!

PS Someone should add IE and Edge screenshots as well. And maybe Epiphany. ;-)

**Comment by lollox** [comment]
Related: https://github.com/OtterBrowser/otter-browser/issues/65#issuecomment-45357300

**Comment by Frenzie** [comment]
Maxthon's menu is the sensible form of what Fx attempts. Thanks for
bringing it to my attention.

**Comment by pierreporte** [comment]
@lollox I added the screenshot of Maxthon and modified the screenshot for Firefox 4 as yours shows an extra feature.

**Comment by Emdek** [comment]
@pierreporte, while I'm planning for long time to add support for such features (#65) I'm still not convinced that we should use them by default (maybe except some minor uses), although we could consider making greater use of them for menu button, since it is kind of crowded...
Anyway, we first need to add support for such controls, I've did some experiments in the past and my conclusion is: `QMenu` sucks, we would either need to replace it with completely custom class or apply loads of hacks, since it is very hard to properly mimic all types of entries using `QActionWidget` and replacing all of them by instances of this class will degrade performance and eat more memory.

**Comment by pierreporte** [comment]
It was recently announced that Firefox will have a brand new design (called “Photon”) for its version 57 due in November 2017. The main menu will change to something like this:

![](http://www.omgubuntu.co.uk/wp-content/uploads/2017/04/firefox-photon-main-menu.jpg)

This is to be used as a reference alongside previous screenshots.

More screenshots here: https://www.soeren-hentzschel.at/firefox/photon-weitere-bilder-zum-neuen-firefox-design/

**Comment by Emdek** [comment]
@pierreporte, appears to be saner that current one.
Assuming that I'm interpreting that correctly. ;-)

**Comment by Frenzie** [comment]
Yeah, that looks pretty good to me. More contrast in the icons (as opposed to slightly darker gray on slightly lighter gray) and a much clearer more or less regular list-type menu.

**Comment by Emdek** [comment]
@Frenzie, for me it's improvement over list of semi random icons with labels in sort of grid, but I'm not 100% sure that it's replacement for that menu and if that menu is what it;s by default nowadays. ;-)

**Comment by Frenzie** [comment]
Right, that's exactly what I mean. :-)

**Comment by cofi89** [comment]
FWIW/IMO, while the new FF menu design is much better and cleaner in the "looks" department, GTK3 Popover/Non-cascading style is a serious drawback and a pain on the usage side of things.  
Seriously, I'm having a hard time understanding this recent trend of making everything "clean" by removing, hiding and obfuscating stuff and navigation...  
  

Taking the proposed FF menu as example, say you wanted to open "Web developer" submenu and instead opened "More". Instead of just moving the mouse ~10px above, as you would in a regular cascading menu, you now have to go all the way up, click the "back" button, then go down again and open "Web Dev".  
Why waste time and more importantly, put unnecessary congitive load on the user,,.? Not to mention, remove the overview of menu structure...?  

Rant aside, I must ask you, please don't do this in Otter! :)

**Comment by Frenzie** [comment]
Well sure, but we're talking about the completely unusable Firefox menu here. Keeping that particular aspect is annoying, but the way it's shown in the screenshot still means a major improvement over the current mess. I suspect they want a consistent experience across all devices, and on tiny mobile screens having menus behave this way constitutes a necessity.

![screenshot_2017-04-10_23-15-38](https://cloud.githubusercontent.com/assets/202757/24883107/d5299184-1e43-11e7-8fe0-39044c2b3f0b.png)
![screenshot_2017-04-10_23-15-50](https://cloud.githubusercontent.com/assets/202757/24883108/d53cb34a-1e43-11e7-9225-cd3fd4df391c.png)

**Comment by cofi89** [comment]
@Frenzie 

No disagreements there! :)  
Except maybe that, considering all the work needed to keep Desktop and Mobile going, having a separate menu styles is not that much of a burden nor a legitimate reason to, well,  force mobile UI on desktop users.

**Comment by Emdek** [comment]
@cofi89, no worries, I'm only considering putting few actions into single menu entry as subactions, mainly zoom related stuff, probably just for menu button anyway.
And so far menu button is not even enabled by default.
Plus we allow to customize it, not sure about others, I doubt that Chrome allows to modify existing entries, but who knows...

**Comment by pierreporte** [comment]
Firefox 57 will have a new main menu alongside the “Photon“ redesign. Here is a screenshot asa reference:

![](https://cdn.ghacks.net/wp-content/uploads/2017/05/firefox.57-photon-design.png)
